src/app/Library/CrudPanel/Traits/ColumnsProtectedMethods.php

```diff
@@ -80,14 +80,55 @@ protected function makeSureColumnHasLabel($column)
      */
     protected function makeSureColumnHasType($column)
     {
-        $could_be_relation = isset($column['entity']) && $column['entity'] !== false;
+        // Do not alter type if it has been set by developer
+        if (isset($column['type'])) {
+            return $column;
+        }
+
+        // Set text as default column type
+        $column['type'] = 'text';
+
+        $could_be_relation = Arr::get($column, 'entity', false) !== false;
 
-        if (! isset($column['type']) && $could_be_relation) {
+        if ($could_be_relation) {
             $column['type'] = 'relationship';
         }
 
-        if (! isset($column['type'])) {
-            $column['type'] = 'text';
+        if (in_array($column['name'], $this->model->getDates())) {
+            $column['type'] = 'datetime';
+        }
+
+        if ($this->model->hasCast($column['name'])) {
+            $attributeType = $this->model->getCasts()[$column['name']];
+
+            switch ($attributeType) {
+                case 'array':
+                case 'encrypted:array':
+                case 'collection':
+                case 'encrypted:collection':
+                case 'json':
+                case 'object':
+                    $column['type'] = 'array';
+                    break;
+                case 'bool':
+                case 'boolean':
+                    $column['type'] = 'check';
+                    break;
+                case 'date':
+                    $column['type'] = 'date';
+                    break;
+                case 'datetime':
+                    $column['type'] = 'datetime';
+                    break;
+                case 'double':
+                case 'float':
+                case 'int':
+                case 'integer':
+                case 'real':
+                case 'timestamp':
+                    $column['type'] = 'number';
+                    break;
+            }
```

Comment:
@jnoordsij thank you so much for you effort here. 

Wouldn't be safer to have a `default` here that returns `text` if we can't match the attribute casting with anything we specify?

Reply:
The default `text` is already set on line 89. Everything after that basically 'overrides' this value if a more suitable type can be found (e.g. relationship check that already existed and new type check added now), but leaves the type the same if no match is found.

Reply:
Indeed, sorry I missed that line somehow and was expecting the fallback at the end.